docs/adr/0005-feature-flags.md

```diff
@@ -0,0 +1,184 @@
+```
+
+### Ownership
+Looking ahead, there may come a time when we need a system that enables the ability to assign ownership of a flag
+
+Feature flags offer a powerful way to directly modify the behavior of production systems, acting as a kill switch. The configuration of flags should undergo the same level of change control as the deployment of production code. Associating explicit ownership with a flag allows for the application of change control policies. For instance, individuals can be restricted from modifying flags they do not own, or approval can be required before making feature-flag configuration changes in the production environment.
+
+### Flag Tagging
+Attaching information like expiration dates and ownership to a flag are specific examples of a more general capability: the ability to tag flags.
+
+Tagging refers to the ability to attach arbitrary key:value pairs to a feature flag. This concept is employed in infrastructure systems like Kubernetes and AWS. It provides a highly adaptable method for users to associate semantic metadata with entities within the system. Tags can be used to track various aspects of a flag, such as its creation date, expiration date, life-cycle status, flag type, ownership, authorized modifier etc.
+
+## Considered Options
+* Environment Variables
+* JSON
+* [Unleash](https://www.getunleash.io/) Open Source Feature Flag (self host)
+* [Unleash](https://www.getunleash.io/) (managed cloud)
+
+## Decision Outcome
+
+TBD
+
+
+## Pros and Cons of the Options <!-- optional -->
+
+### Environment Variable
+
+```
+ENABLED_FEATURE_FLAGS = "premium_subplat_email, free_breaches_hibp"
+
+function isFlagEnabled(flag: string): boolean {
+  return (process.env.ENABLED_FEATURE_FLAGS ?? '').split(',').includes(flag);
+}
+
+const confettiScript = isFlagEnabled('party-mode')
+  ? '<script src="confetti.js"></script>'
+  : "";
+```
+
+* Good, because it's the easiest/quickest way to implement feature flags
+* Good, because the changes can be easily tested in separate environments, programmatically
+* Good, because the changes can be scalable (by simply adding and subtracting from the env var array)
+* Good, because it's easy to maintain and it's centralized
+* Good, because it's easy to learn and understand
+* Good, having SRE involved adds another level of sanity check
+* Bad, because we will become dependent on SRE to make any changes in stage and production environment. However, it's not too bad because we will likely have to involve SRE when releasing to production anyways, and during release is when we will be deciding and modifying the environment variables. 
+* Bad, because the "implementation" is limited to the current repo. In a world where we have multiple repos serving different parts of the app, there's no easy way to share
+
+### JSON
```

Comment:
What's unclear to me from this description is where this JSON would live? In an env var, or in a file in the repo? If the latter, how easy would it be to have different flags per environment? And then I'd also list as a downside that flipping a flag (e.g. if it introduces a bug that we quickly want to back out) would require a new hotfix release and deploy, which is probably riskier than just redeploying the same code with a different environment variable.

Reply:
Oh, I suppose the `environments` property adapts it to different environments. Shooting from the hip: maybe we could even add an `envVar` property that looks for a particular value in a feature-flag-specific env var, and then enable (or disable?) it when that value is present? Then we'd have the best of both worlds.

Reply:
Hmm good question. I think if it's JSON, the assumption is that the JSON would live in the code.

We definitely need to think about the pros and cons of these two approaches a bit more...
cc @rhelmer @maxxcrawford @flozia @codemist 

1. `env var` approach seems pretty straightforward, but it's actually not in practice:
    1. we actually have a hard time as devs (even SREs) to have a holistic picture of what are the env vars that are currently in place for stage and production environment. These are values that are buried deep inside of config maps and yaml files in moz services repo, which I don't know how many of our team has access
    2. changing `stage` and `prod` env vars require SRE approval / restart the server
2. let's think about the deployment of `json` approach
    1. changes to `stage` can be done by devs via our normal deployment process
    2. changes to `prod` requires SRE approval / restart
    3. I agree with you that the changes would be riskier since it'll be code changes -- but I guess if we do feature flags correctly, feature flags are meant for reducing this exact risk?
    4. we (at least as devs) will have a better view of what's in dev/stage/prod, re the first point under `env var`
    5. Not to mention all the other config benefits from having a json structure already in the doc
    

Separately, I'm thinking maybe the below format makes more sense, a bit more dup in data, but faster retrieval..:
```json
{
  "dev": {
    "name": "Premium Subplat Email", 
    "description": "description",
    "key": "premium_subplat_email",
    "createdAt": "",
    "expiresAt": "",
    ...
  }
  "stage":
  "prod":
}
```

Reply:
Regarding

> we actually have a hard time as devs (even SREs) to have a holistic picture of what are the env vars that are currently in place for stage and production environment. 

on Relay, we just expose the feature flag's env var's value for inspection ([at `/runtime_data/`](https://relay.firefox.com/api/v1/runtime_data)).

As for

> I agree with you that the changes would be riskier since it'll be code changes -- but I guess if we do feature flags correctly, feature flags are meant for reducing this exact risk?

I guess the question is if we're OK with an always-deployable `main` (in which case we could just push the latest `main` through to production with the feature flag flipped), or if we still want to make sure that commits since the latest deployments are not going to prod. If the former, the risk is indeed not increased, but if the latter, I'd at least mention it as a downside, though I wouldn't necessarily consider it reason not to go ahead with it.

Agreed with the rest of your analysis and your proposal.

Reply:
@Vinnl Good points. 

Some of these feature flag discussions inevitably have to turn into "deployment process" discussions. For Monitor, so far, we haven't gone beyond the "always deployable main" model yet. We did have some tech meetings about potentially adopting a release branch strategy.. which escalates to having a dedicated "release engineer" or someone who is going to be in charge of cherry picking at the time of release.. which escalates to longer release cycles. I think for the size that we are at, the "always deployable main" strategy still seems to work, though I do have to acknowledge that the more people we have, the more chaotic `main` becomes

Reply:
Oh heh, I'd much prefer the always-deployable `main`, I just wasn't sure if we were there yet :sweat_smile: